veps/sig-network/link-state-proposal.md

```diff
@@ -0,0 +1,217 @@
+# VEP #3: Adding Link State Management for vNICs
+
+## Release Signoff Checklist
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [X] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] (not the initial VEP PR)
+
+## Overview
+
+The purpose of this enhancement is to introduce support for link state management (up/down) in KubeVirt. This feature
+will enable VM owners to dynamically control the link state. It aligns KubeVirt with traditional virtualization platforms
+and enhances its utility in environments requiring precise network state control.
+
+By introducing link state management, KubeVirt will:
+- Provide VM-level controls to configure the link state of individual interfaces.
+- Report link state info via the VirtualMachineInterface status.
+
+This proposal outlines the necessary changes to KubeVirt, including API changes.
+
+## Motivation
+
+Currently, KubeVirt does not support modifying or reporting the link state of virtual network interfaces.
+This limitation can lead to challenges in implementing advanced networking scenarios, such as:
+- Simulating link failures for testing and debugging.
+- Managing multi-network setups, where some interfaces need to be intentionally disabled.
+
+## Goals
+
+1. Allow users to configure the link state (up/down) of individual interfaces through the KubeVirt API.
+2. Enable link state configuration at both VM creation and runtime.
+3. Report the current domain link state of each VM interface.
+
+## Non Goals
+
+1. Affect network plumbing on the host - the setup process will remain unchanged.
+2. Report the link state of the underlying network provider.
+3. Report the link state of interfaces inside the guest operating system.
+
+## User Stories
+
+As a VM owner I would like to be able to:
+
+- Start the VM with one or more interfaces with their link set to `down`.
+- Toggle the link state of one or more interfaces of a running VM.
+- Hot plug an interface with a link state set to `down`.
+- Control the link state of interfaces connected to primary and secondary networks.
```

Comment:
To follow-up on our offline discussion - I do see a value in having it on the primary network from the start. The motivation to isolate a running VM from the rest of the network applies for both primary and secondary. If there are any strong specific concerns about link-state control causing issues to primary CNIs, please raise them @maiqueb @EdDev

Reply:
On the primary network, we heavily depend on dynamic IP configuration served by Kubevirt (in the virt-launcher).
Moreover, IPv6 full support requires cloud-init intervention.

Unlike the secondary networks which we depend much more on the VM owner and operator, on the primary we have more responsibility.
As I see it, that increases the risk and with it the need to cover much more scenarios than on secondary networks.
Said that, based on preliminary checks, we can decide to take the risk and warn about the implications of using it on the primary network without blocking it.

Reply:
Could you share more about those two examples? I don't understand how IP configuration serving or IPv6 is affected by the link state.

Reply:
For dynamic configuration (e.g. DHCP) the client needs to successfully re-request to be assign an IP.
The server side (Kubevirt) needs to successfully answer it.
Or some network provider thing (OVN-K?) needs to answer it.

Same goes with configuration coming from cloud-init (e.g. for IPv6), it needs to be removable and it presents different scenarios: When link was down on boot and came up later, when link went down and up while VM is running, etc...

Reply:
How are those scenarios different from soft reboot, which we do support today with all network options?

Reply:
Please resolve this thread.